File: 05_practice/src/com/gomboclabs/codespirit/practice/basic/class01/java/Code02_BubbleSort.java
package gomboclabs.codespirit.practice.basic.class01.java;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.util.Arrays;

/**
 * @author xucg
 * @version 1.0
 * Created on 2023/9/8 - 09 - 08
 * description:
 */
public class Code02_BubbleSort {

    private static final Logger log = LogManager.getLogger(Code02_BubbleSort.class);

    // execution time: 4344
    // public static void bubbleSortNotOptimized(int[] arr) {
    //     for (int i = 0; i < arr.length; i++) {
    //         for (int j = arr.length - 1; j > i; j--) {
    //             if (arr[j] < arr[j - 1]) {
    //                 swap(arr, j, j - 1);
    //             }
    //         }
    //     }
    // }

    // 优化后的算法，记录上一次冒泡是否交换，判断后面是否还进行冒泡
    // execution time: 4206 4069ms
    public static void bubbleSort(int[] arr) {
        boolean swapped;
        for (int i = 0; i < arr.length; i++) {
            swapped = false;
            for (int j = arr.length - 1; j > i; j--) {
                if (arr[j] < arr[j - 1]) {
                    swap(arr, j, j - 1);
                    swapped = true;
                }
            }
            if (!swapped) {
                log.debug("No more swaps needed. Exiting loop.");
                break;
            }
        }
    }

    // 交换arr的i和j位置上的值
    public static void swap(int[] arr, int i, int j) {
        arr[i] = arr[i] ^ arr[j];
        arr[j] = arr[i] ^ arr[j];
        arr[i] = arr[i] ^ arr[j];
    }

    // for test
    public static void comparator(int[] arr) {
        Arrays.sort(arr);
    }

    // for test
    public static int[] generateRandomArray(int maxSize, int maxValue) {
        int[] arr = new int[(int) ((maxSize + 1) * Math.random())];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = (int) ((maxValue + 1) * Math.random()) - (int) (maxValue * Math.random());
        }
        return arr;
    }

    // for test
    public static int[] copyArray(int[] arr) {
        if (arr == null) {
            return null;
        }
        int[] res = new int[arr.length];
        for (int i = 0; i < arr.length; i++) {
            res[i] = arr[i];
        }
        return res;
    }

    // for test
    public static boolean isEqual(int[] arr1, int[] arr2) {
        if ((arr1 == null && arr2 != null) || (arr1 != null && arr2 == null)) {
            return false;
        }
        if (arr1 == null && arr2 == null) {
            return true;
        }
        if (arr1.length != arr2.length) {
            return false;
        }
        for (int i = 0; i < arr1.length; i++) {
            if (arr1[i] != arr2[i]) {
                return false;
            }
        }
        return true;
    }

    // for test
    public static void printArray(int[] arr) {
        if (arr == null) {
            return;
        }
        for (int i = 0; i < arr.length; i++) {
            System.out.print(arr[i] + " ");
        }
        System.out.println();
    }

    // for test
    public static void main(String[] args) {
        Long startTime = System.currentTimeMillis();
        int testTime = 500000;
        int maxSize = 100;
        int maxValue = 100;
        boolean succeed = true;
        for (int i = 0; i < testTime; i++) {
            int[] arr1 = generateRandomArray(maxSize, maxValue);
            int[] arr2 = copyArray(arr1);
            bubbleSort(arr1);
            comparator(arr2);
            if (!isEqual(arr1, arr2)) {
                succeed = false;
                break;
            }
        }
        System.out.println(succeed ? "Nice!" : "Fucking fucked!");
        Long endTime = System.currentTimeMillis();
        System.out.println("execution time: " + (endTime - startTime) + "ms");
        // int[] arr = generateRandomArray(maxSize, maxValue);
        // printArray(arr);
        // bubbleSort(arr);
        // printArray(arr);
    }
}

File: 05_practice/src/com/gomboclabs/codespirit/practice/basic/class01/js/Code02_BubbleSort.js
// 练手
function bubbleSort(arr) {
    for (let i = 0; i < arr.length; i++) {
        let isSwapped = false;
        for (let j = arr.length - 1; j > i ; j--) {
            if (arr[j] < arr[j-1]) {
                isSwapped = true;
                swap(arr, j, j-1);
            }
        }
        if (!isSwapped) {
            break;
        }
    }
}

function swap(arr, i, j) {
    const temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

function comparator(arr) {
    arr.sort((a, b) => a - b);
}

function generateRandomArray(maxSize, maxValue) {
    const arr = [];
    for (let i = 0; i < Math.floor(Math.random() * maxSize); i++) {
        arr.push(Math.floor(Math.random() * (2 * maxValue + 1)) - maxValue);
    }
    return arr;
}

function copyArray(arr) {
    return arr.slice();
}

function isEqual(arr1, arr2) {
    if (arr1 === null && arr2 === null) {
        return true;
    }
    if (arr1 === null || arr2 === null || arr1.length !== arr2.length) {
        return false;
    }
    for (let i = 0; i < arr1.length; i++) {
        if (arr1[i] !== arr2[i]) {
            return false;
        }
    }
    return true;
}

function printArray(arr) {
    console.log(arr.join(' '));
}

function main() {
    console.time("executionTime");
    const testTime = 500000;
    const maxSize = 100;
    const maxValue = 100;
    let succeed = true;
    for (let i = 0; i < testTime; i++) {
        const arr1 = generateRandomArray(maxSize, maxValue);
        const arr2 = copyArray(arr1);
        bubbleSort(arr1);
        comparator(arr2);
        if (!isEqual(arr1, arr2)) {
            succeed = false;
            printArray(arr1);
            printArray(arr2);
            break;
        }
    }
    console.log(succeed ? 'Nice!' : 'Fucking fucked!');
    console.timeEnd("executionTime");
}

main();